src/Microsoft.AspNetCore.Mvc.Abstractions/ModelBinding/ModelStateDictionary.cs

```diff
@@ -741,7 +910,47 @@ public PrefixEnumerable FindKeysWithPrefix(string prefix)
             return new PrefixEnumerable(this, prefix);
         }
 
-        public struct PrefixEnumerable : IEnumerable<KeyValuePair<string, ModelStateEntry>>
+        private class ModelStateNode : ModelStateEntry
+        {
+            public ModelStateNode(ModelStateNode parent, StringSegment key)
+            {
+                Debug.Assert(key != null);
+                Parent = parent;
+                Key = key;
+            }
+
+            public ModelStateNode Parent { get; }
+
+            public List<KeyValuePair<StringSegment, ModelStateNode>> Children { get; set; }
+
+            public bool Visible { get; set; }
+
+            public StringSegment Key { get; }
```

Comment:
Could we represent it internally as `StringSegment`? The idea being we wouldn't have to create strings for intermediate (not `Visible`) nodes?